### ISSUE TITLE
DeDRM plugin failed to install

### ISSUE THREAD

**Issue by hallofthemtking** [apprenticeharper/DeDRM_tools#1019]
Installing DeDRM plugin to Calibre (latest version) Get the following:
calibre, version 4.11.2
ERROR: Unhandled exception: <b>InvalidPlugin</b>:The plugin in u'/Users/bradcrandell/Downloads/DeDRM_tools_6.7.0.zip' is invalid. It does not contain a top-level __init__.py file

Python 2.7.16
OSX: ('10.15.3', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: Gather KFX-ZIP (from KFX Input) (1, 9, 0) && Package KFX (from KFX Input) (1, 9, 0) && KFX metadata reader (from KFX Input) (1, 9, 0) && KFX Input (1, 9, 0) && Set KFX metadata (from KFX Output) (1, 23, 0) && KFX Output (1, 23, 0)
Traceback (most recent call last):
  File "site-packages/calibre/gui2/preferences/plugins.py", line 319, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 472, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 61, in load_plugin
  File "site-packages/calibre/customize/zipplugin.py", line 194, in load
  File "site-packages/calibre/customize/zipplugin.py", line 292, in _locate_code
InvalidPlugin: The plugin in u'/Users/bradcrandell/Downloads/DeDRM_tools_6.7.0.zip' is invalid. It does not contain a top-level __init__.py file

**Comment by ElleKayEm** [comment]
You're trying to install the whole archive.  Unzip DeDRM_tools_6.7.0.zip and load DeDRM_Plugin.zip into calibre.

**Comment by hallofthemtking** [comment]
Thanks, got it.  Won’t remove the drm of Kindle books however.

**Comment by ElleKayEm** [comment]
Post a log:

https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md#i-cannot-solve-my-problem-with-the-dedrm-plugin-and-now-i-need-to-post-a-log-how-do-i-do-that

**Comment by ealcaniz** [comment]
I have the same issue but with Fedora 31

calibre 4.10.1  embedded-python: False is64bit: True
Linux-5.5.7-200.fc31.x86_64-x86_64-with-fedora-31-Thirty_One Linux ('64bit', 'ELF')
('Linux', '5.5.7-200.fc31.x86_64', '#1 SMP Fri Feb 28 17:18:37 UTC 2020')
Python 3.7.6
Linux: ('Fedora', '31', 'Thirty One')
Interface language: en_GB
Successfully initialized third party plugins: 
Traceback (most recent call last):
  File "/usr/lib64/calibre/calibre/gui2/preferences/plugins.py", line 318, in add_plugin
    plugin = add_plugin(path)
  File "/usr/lib64/calibre/calibre/customize/ui.py", line 472, in add_plugin
    plugin = load_plugin(path_to_zip_file)
  File "/usr/lib64/calibre/calibre/customize/ui.py", line 61, in load_plugin
    return loader.load(path_to_zip_file)
  File "/usr/lib64/calibre/calibre/customize/zipplugin.py", line 200, in load
    reload(m)
  File "/usr/lib64/calibre/polyglot/builtins.py", line 112, in reload
    return importlib.reload(module)
  File "/usr/lib64/python3.7/importlib/__init__.py", line 169, in reload
    _bootstrap._exec(spec, module)
  File "<frozen importlib._bootstrap>", line 628, in _exec
  File "/usr/lib64/calibre/calibre/customize/zipplugin.py", line 180, in load_module
    import_name), 'exec', dont_inherit=True)
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
                                                                    ^
SyntaxError: invalid syntax

**Comment by ealcaniz** [comment]
[ealcaniz@ealcaniz customize]$ ll
total 180
-rw-r--r--. 1 root root 66529 Feb 15 22:37 builtins.py
-rw-r--r--. 1 root root 14166 Feb 15 22:37 conversion.py
-rw-r--r--. 1 root root 27161 Feb 15 22:37 __init__.py
-rw-r--r--. 1 root root 26582 Feb 15 22:37 profiles.py
-rw-r--r--. 1 root root 25885 Feb 15 22:37 ui.py
-rw-r--r--. 1 root root 11705 Feb 15 22:37 zipplugin.py

[ealcaniz@ealcaniz customize]$ pwd
/usr/lib64/calibre/calibre/customize
[ealcaniz@ealcaniz customize]$

**Comment by ElleKayEm** [comment]
See #986 and #928.

**Comment by racoco** [comment]
I am running FC-31 and Calibre 4.12.0.
I also experienced the same issue as ealcaniz when attempting to install the DeDRM_plugin.zip (it is v6.6.3). tho for me the exception got thrown on line 167 of the __init__.py file inside the zip bundle.

[But I was able to successfully install the other recommended plugin: kfx-input.zip]

Looking at that line of code, I commented it out (it's just a print statement after all) to see what would happen. Well, a similar print statement on line 179 threw an exception. Unfortunately, it was inside an exception handling block so clearly I was screwed. I could not figure out what was wrong, so I googled this issue and found this and other pages.

Reading it all, it seemed clear that the issue had to do with Calibre and Python 3 incompatibilities and that there was nothing to be done until the Calibre folks addressed the issue.

Out of curiosity, I imported the AZW4 Kindle Fire file for the Kindle eBook I recently bought. The import succeeded w/o incident, so I took the next step of converting it to PDF expecting to see decryption errors of some type.

WTF!!!  It worked!!!

I have no idea why. Maybe something in the kfx-input plugin? Maybe the Kindle file I purchased was not encrypted? Dunno, but I got what I wanted and don't wish to make a career over the DeDRM plugin issue.

Has Amazon stopped encrypting their Kindle books?
Everything I see online wrt turning Amazon Kindle content to another format tals about using Calibre and DeDRM. I didn't seem to need DeDRM! 

Just thought I'd pass this on.
Would be happy to hear an explanation if someone has one.

Rich

**Comment by ElleKayEm** [comment]
If it was an AZW4 from a Kindle Fire, it must have already been DRM-free.  Because DeDRM can't decrypt them (only from E-ink Kindles and Kindle for PC/Mac).  KFX input only helps with KFX format.

**Comment by racoco** [comment]
ElleKayEm - thank you!

**Comment by vremja** [comment]
I'm running Calibre 5.0.1 on MacOS Mojave 10.14.6, followed the instructions for installing the plugin and it gives me this error:

calibre, version 5.0.1
ERRORE: Eccezione non gestita: <b>SyntaxError</b>:invalid syntax (calibre_plugins.dedrm.__init__, line 168)

calibre 5.0.1  embedded-python: True is64bit: True
macOS-10.14.6-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '18.7.0', 'Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.14.6', ('', '', ''), 'x86_64')
Interface language: it
Successfully initialized third party plugins: 
Traceback (most recent call last):
  File "site-packages/calibre/gui2/preferences/plugins.py", line 317, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 472, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 61, in load_plugin
  File "site-packages/calibre/customize/zipplugin.py", line 200, in load
  File "site-packages/polyglot/builtins.py", line 123, in reload
  File "lib/python3.8/importlib/__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 602, in _exec
  File "site-packages/calibre/customize/zipplugin.py", line 179, in load_module
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
          ^
SyntaxError: invalid syntax

how do I resolve the problem?

**Comment by vremja** [comment]
I'm running Calibre 5.0.1 on MacOS Mojave 10.14.6, followed the instructions for installing the plugin and it gives me this error:

calibre, version 5.0.1
ERRORE: Eccezione non gestita: <b>SyntaxError</b>:invalid syntax (calibre_plugins.dedrm.__init__, line 168)

calibre 5.0.1  embedded-python: True is64bit: True
macOS-10.14.6-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '18.7.0', 'Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.14.6', ('', '', ''), 'x86_64')
Interface language: it
Successfully initialized third party plugins: 
Traceback (most recent call last):
  File "site-packages/calibre/gui2/preferences/plugins.py", line 317, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 472, in add_plugin
  File "site-packages/calibre/customize/ui.py", line 61, in load_plugin
  File "site-packages/calibre/customize/zipplugin.py", line 200, in load
  File "site-packages/polyglot/builtins.py", line 123, in reload
  File "lib/python3.8/importlib/__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 602, in _exec
  File "site-packages/calibre/customize/zipplugin.py", line 179, in load_module
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
          ^
SyntaxError: invalid syntax
How do I solve the problem?

**Comment by net-buoy** [comment]
Plug-in does not run with Python 3 yet

**Comment by ElleKayEm** [comment]
@vremja 
For now stick with [calibre 4.23](https://download.calibre-ebook.com/4.23.0/) -- the last Python 2 version.

**Comment by JulieUK** [comment]
> For now stick with [calibre 4.23](https://download.calibre-ebook.com/4.23.0/) -- the last Python 2 version.

Thanks for that helpful piece of info!  This is the only way I have to get digital library books onto an old Kindle keyboard for my son to read, and e-books are all we can get hold of right now.  I recently upgraded to Calibre 5.0 on Ubuntu and didn't have any problems with the previously installed DeDRM plugin.  Then I got a pop-up inviting me to upgrade to 5.2, and that conclusively broke the DeDRM functionality.  I tried reverting to 5.0 and it didn't work any more either.  After I tried 3.42 which didn't want to work properly with the plugin either, I was about to give up when I spotted your post.

**Comment by JulieUK** [comment]
> For now stick with [calibre 4.23](https://download.calibre-ebook.com/4.23.0/) -- the last Python 2 version.

Thanks for that helpful piece of info!  This is the only way I have to get digital library books onto an old Kindle keyboard for my son to read, and e-books are all we can get hold of right now.  I recently upgraded to Calibre 5.0 on Ubuntu and didn't have any problems with the previously installed DeDRM plugin.  Then I got a pop-up inviting me to upgrade to 5.2, and that conclusively broke the DeDRM functionality.  I tried reverting to 5.0 and it didn't work any more either.  After I tried 3.42 which didn't want to work properly with the plugin either, I was about to give up when I spotted your post.
